### PR TITLE
feat: add organization alias course wide

### PR DESCRIPTION
**Pull request by uzairr**
## Description:
To create a hierarchy inside an organization in a way that the sub-departments of it are also reflected on course about page.

## Linked Ticket:
for more details please visit: [PROD-2719](https://openedx.atlassian.net/browse/PROD-2719)

### CODE REVIEW

File: course_discovery/apps/course_metadata/models.py [hunk 1]
```diff
@@ -800,6 +800,13 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
         related_name='related_courses',
         default=None, null=True, blank=True
     )
+    logo_image_override = models.ImageField(
+        upload_to=UploadToFieldNamePath(populate_from='uuid', path='media/course/org/logo_override'),
+        blank=True,
+        null=True,
+        validators=[FileExtensionValidator(['png'])]
+    )
+    organization_name_override = models.CharField(max_length=255, default=None, blank=True)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: default=None seems weird.

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -831,6 +831,19 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
...
+        null=True,
+        variations={
+            'large': (256, 256),
+            'medium': (128, 128),
+            'small': (64, 64),
+            'x-small': (32, 32),
+        },
+        help_text=_('Please provide an image file with transparent background'),
+    )
+    organization_name_override = models.CharField(max_length=255, blank=True)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: well, it is not name. short code to be precise. let's be consistent with the naming conventions here.

File: course_discovery/apps/api/serializers.py [hunk 3]
```diff
@@ -1117,6 +1117,8 @@ class CourseSerializer(TaggitSerializer, MinimalCourseSerializer):
                                                        read_serializer=CollaboratorSerializer())
     skill_names = serializers.SerializerMethodField()
     skills = serializers.SerializerMethodField()
+    organization_name_override = serializers.CharField(required=False)
+    logo_image_override = ImageField(read_only=True, source='logo_image_override_url')
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: please update fied names in model and serializer:
- It is organization short code override, not name
- What logo override? Please specify that.

Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: Why is logo field read_only? How will it work with API during making non-GET operations?

**Review by DawoudSheraz — COMMENTED** [review]
- Missing unit tets
- Field name updates

File: course_discovery/apps/api/serializers.py [hunk 4]
```diff
@@ -1117,6 +1117,8 @@ class CourseSerializer(TaggitSerializer, MinimalCourseSerializer):
                                                        read_serializer=CollaboratorSerializer())
     skill_names = serializers.SerializerMethodField()
     skills = serializers.SerializerMethodField()
+    organization_shortcode = serializers.CharField(required=False)
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: nit: organization_short_code_override

File: course_discovery/apps/api/serializers.py [hunk 5]
```diff
@@ -1117,6 +1117,8 @@ class CourseSerializer(TaggitSerializer, MinimalCourseSerializer):
...
+    org_logo_override = ImageField(required=False, source='org_logo_override_url')
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: nit: organization_logo_override

File: course_discovery/apps/course_metadata/models.py [hunk 6]
```diff
@@ -866,6 +866,19 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
...
+        null=True,
+        variations={
+            'large': (256, 256),
+            'medium': (128, 128),
+            'small': (64, 64),
+            'x-small': (32, 32),
+        },
+        help_text=_('Please provide an image file with transparent background'),
+    )
+    organization_shortcode = models.CharField(max_length=255, blank=True)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: why is null=True not added here?

Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: field name nits here too

File: course_discovery/apps/course_metadata/models.py [hunk 7]
```diff
@@ -866,6 +866,19 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
     salesforce_id = models.CharField(max_length=255, null=True, blank=True)  # Course__c in Salesforce
     salesforce_case_id = models.CharField(max_length=255, null=True, blank=True)  # Case in Salesforce
     type = models.ForeignKey(CourseType, models.CASCADE, null=True)  # while null IS True, it should always be set
+    org_logo_override = StdImageField(
+        upload_to=UploadToFieldNamePath(populate_from='name', path='media/course/org/logo_override/'),
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: what does `name` signify here?

File: course_discovery/apps/api/tests/test_serializers.py [hunk 8]
```diff
@@ -249,6 +251,28 @@ def test_draft_and_official(self):
...
+        course_draft = CourseFactory(draft=True)
+        draft_course_run = CourseRunFactory(draft=True, course=course_draft)
+        course_draft.canonical_course_run = draft_course_run
+        course_draft.save()
+
+        course = CourseFactory(draft=False, draft_version_id=course_draft.id)
+        course_run = CourseRunFactory(draft=False, course=course, draft_version_id=draft_course_run.id)
+        course.canonical_course_run = course_run
+        course.organization_shortcode = 'testorg_overriden'
+        course.org_logo_override = course.image
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/tests/test_serializers.py]: can we have another image here, even though it is test, just to keep the differentation?

File: course_discovery/apps/api/tests/test_serializers.py [hunk 9]
```diff
@@ -249,6 +251,28 @@ def test_draft_and_official(self):
...
+        course.save()
+
+        serializer = self.serializer_class(course, context={'request': request, 'exclude_utm': 1, 'editable': 1})
+
+        assert serializer.data['organization_shortcode'] is not None
+        assert serializer.data['organization_shortcode'] == 'testorg_overriden'
+
+        assert serializer.data['org_logo_override'] is not None
+        assert serializer.data['org_logo_override'].get('src') is not None
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/tests/test_serializers.py]: assert against the src value

File: course_discovery/apps/api/serializers.py [hunk 10]
```diff
@@ -1162,7 +1164,7 @@ class Meta(MinimalCourseSerializer.Meta):
             'extra_description', 'additional_information', 'additional_metadata', 'faq', 'learner_testimonials',
             'enrollment_count', 'recent_enrollment_count', 'topics', 'partner', 'key_for_reruns', 'url_slug',
             'url_slug_history', 'url_redirects', 'course_run_statuses', 'editors', 'collaborators', 'skill_names',
-            'skills',
+            'skills', 'organization_shortcode', 'org_logo_override',
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: does including url here auto-downloads the object when saving?

**Review by DawoudSheraz — COMMENTED** [review]
Nits and small updates

File: course_discovery/apps/course_metadata/models.py (hunk 6 above)
Comment (uzairr) [reply course_discovery/apps/course_metadata/models.py]: making null and blank to `True` would create the two different situations for empty string so enabling one of them will avoid that scenario.

File: course_discovery/apps/course_metadata/models.py (hunk 7 above)
Comment (uzairr) [reply course_discovery/apps/course_metadata/models.py]: it is using `name` the same way as `uuid` for creating uniqueness.

File: course_discovery/apps/api/serializers.py (hunk 4 above)
Comment (azanbinzahid) [reply course_discovery/apps/api/serializers.py]: Do we need to add this explicitly in serializer? take an example of `full_description` field.

File: course_discovery/apps/course_metadata/models.py (hunk 6 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/models.py]: I guess it will be `null=False, blank=True` now?

**Review by azanbinzahid — CHANGES_REQUESTED** [review]
Please add views tests as well to test update API call via publisher. Precisely here: `course_discovery/apps/api/v1/tests/test_views/test_courses.py `

File: course_discovery/apps/course_metadata/tests/factories.py [hunk 11]
```diff
@@ -244,6 +244,7 @@ class CourseFactory(SalesforceRecordFactory):
     syllabus_raw = FuzzyText()
     outcome = FuzzyText()
     image = factory.django.ImageField()
+    organization_logo_override = factory.django.ImageField()
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/tests/factories.py]: Do we also need `organization_short_code_override` in factory?

**Review by azanbinzahid — COMMENTED** [review]
Looks good, just missing api/view tests. They will verify if API updates are working as expected or not.

File: course_discovery/apps/course_metadata/tests/factories.py (hunk 11 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/tests/factories.py]: +1 here, this should be added in factory too.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py [hunk 12]
```diff
@@ -1019,6 +1019,22 @@ def test_update_with_level_type(self):
...
+                'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNk+A8AAQUBAScY'
+                '42YAAAAASUVORK5CYII=',
+            'organization_short_code_override': 'test_org'
+        }
+        url = reverse('api:v1:course-detail', kwargs={'key': course.uuid})
+        response = self.client.patch(url, payload, format='json')
+        assert response.status_code == 200
+        course = Course.everything.get(uuid=course.uuid, draft=True)
+        self.assertEqual(self.serialize_course(course)['organization_short_code_override'], 'test_org')
+        self.assertDictEqual(response.data, self.serialize_course(course))
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_courses.py]: serialize_course call can be moved above and stored in a variable. Also, add an assertion explicitly against logo override

**Review by DawoudSheraz — COMMENTED** [review]
going to test on local before the approval

File: course_discovery/apps/course_metadata/models.py [hunk 13]
```diff
@@ -889,6 +889,19 @@ class Course(DraftModelMixin, PkSearchableMixin, CachedMixin, TimeStampedModel):
...
+    organization_logo_override = StdImageField(
+        upload_to=UploadToFieldNamePath(populate_from='uuid', path='media/course/org/logo_override/'),
+        blank=True,
+        null=True,
+        variations={
+            'large': (256, 256),
+            'medium': (128, 128),
+            'small': (64, 64),
+            'x-small': (32, 32),
+        },
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: The logo field in Organization model does not have any variations defined there. We can keep the behavior consistent here, unless there is a reason for this.

**Review by DawoudSheraz — APPROVED** [review]
Tested locally on Firebox using Edit request feature to patch the data. The changes are working.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py [hunk 14]
```diff
@@ -1019,6 +1019,22 @@ def test_update_with_level_type(self):
         course = Course.everything.get(uuid=self.course.uuid, draft=True)
         assert course.level_type == beginner
 
+    @responses.activate
+    def test_override_with_shortcode_and_logo_image(self):
+        course = CourseFactory()
```
Comment (azanbinzahid) [inline course_discovery/apps/api/v1/tests/test_views/test_courses.py:1024]: If you are making a course from factory, you may already have these fields populated from factory and this patch req will update those.

**Review by azanbinzahid — APPROVED** [review]

**Review by DawoudSheraz — APPROVED** [review]

**Review by azanbinzahid — APPROVED** [review]

**Review by waheedahmed — APPROVED** [review]

**Pull request merged**